### PR TITLE
[SPARK-45699][BUILD][CORE][SQL][SS][CONNECT][MLLIB][ML][DSTREAM][GRAPHX][K8S][UI] Fixing all compilation warnings related to widening conversions

### PR DESCRIPTION
**Pull request by hannahkamundson**
### What changes were proposed in this pull request?

1. Change the silencing of the widening conversion compilation warnings in the parent `pom.xml` and `SparkBuild` to throw an error
2. All widening conversion compilation warnings were removed. This almost exclusively involved adding `.toDouble` to longs. However, it also involved some `.toFloat` on ints and longs.

### Why are the changes needed?
It allows us to upgrade to Scala 2.13 without adding a bunch of compilation issues. This is removing the following compilation error
```shell
[error] /Users/yangjie01/SourceCode/git/spark-mine-sbt/core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:1207:60: Widening conversion from Long to Double is deprecated because it loses precision. Write `.toDouble` instead. [quickfixable]
[error] Applicable -Wconf / @nowarn filters for this fatal warning: msg=<part of the message>, cat=deprecation, site=org.apache.spark.scheduler.TaskSetManager.checkSpeculatableTasks
[error]       foundTasks = checkAndSubmitSpeculatableTasks(timeMs, threshold, customizedThreshold = true)
```

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
No tests were added. 
For every profile (including the base profile), I ran `mvn clean compile test-compile`. I then `grep`ed any lines that had the word `Wide` in it. I determined I was done when no output remained.

Here is a script of what was run:
```shell
mvn clean compile test-compile |& tee output.txt
cat output.txt | grep .*Wide.* |& tee output-widening.txt
mvn clean compile test-compile -Pspark-ganglia-lgpl |& tee output-spark-ganglia-lgpl.txt
cat output-spark-ganglia-lgpl.txt | grep .*Wide.* |& tee output-spark-ganglia-lgpl-widening.txt
mvn clean compile test-compile -Pkinesis-asl |& tee output-kinesis-asl.txt
cat output-kinesis-asl.txt | grep .*Wide.* |& tee output-kinesis-asl-widening.txt
mvn clean compile test-compile -Pdocker-integration-tests |& tee output-docker-integration-tests.txt
cat output-docker-integration-tests.txt | grep .*Wide.* |& tee output-docker-integration-tests-widening.txt
mvn clean compile test-compile -Pyarn \& tee output-yarn.txt
cat output-yarn.txt | grep .*Wide.* |& tee output-yarn-widening.txt
mvn clean compile test-compile -Pkubernetes |& tee output-kubernetes.txt
cat output-kubernetes.txt | grep .*Wide.* |& tee output-kubernetes-widening.txt
mvn clean compile test-compile -Pkubernetes-integration-tests |& tee output-kubernetes-integration-tests.txt
cat output-integration-tests.txt | grep .*Wide.* |& tee output-integration-tests-widening.txt
mvn clean compile test-compile -Phive-thriftserver |& tee output-hive-thriftserver.txt
cat output-thriftserver.txt | grep .*Wide.* |& tee output-thriftserver-widening.txt
mvn clean compile test-compile -Phadoop-cloud |& tee output-hadoop-cloud.txt
cat output-hadoop-cloud.txt | grep .*Wide.* |& tee output-hadoop-cloud-widening.txt
```

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Comment by hannahkamundson** [comment]
@LuciferYang My CICD build is failing because of scalastyle. Could you point me towards a solution to fix this?

File: core/src/main/scala/org/apache/spark/resource/ResourceUtils.scala
Comment (LuciferYang) [inline core/src/main/scala/org/apache/spark/resource/ResourceUtils.scala:479]: Before this pr, the result here was a Long value from `execAmount * numParts / taskAmount`. Is there a need to perform `Math.floor` on it? @tgravescs Do you have time to help check this logic? Thanks

File: core/src/test/scala/org/apache/spark/benchmark/Benchmark.scala
Comment (LuciferYang) [inline core/src/test/scala/org/apache/spark/benchmark/Benchmark.scala:166]: how about just make `avg` as a `Double`?

File: core/src/test/scala/org/apache/spark/deploy/history/EventLogTestHelper.scala
Comment (LuciferYang) [inline core/src/test/scala/org/apache/spark/deploy/history/EventLogTestHelper.scala:59]: This is the third case related to `Math.floor`. 
`desiredSize / stringLen` is also a long value. Is it necessary to perform `Math.floor` here?

 @HeartSaVioR  Do you have time to help check this logic? Thanks

File: graphx/src/test/scala/org/apache/spark/graphx/lib/PageRankSuite.scala
Comment (LuciferYang) [inline graphx/src/test/scala/org/apache/spark/graphx/lib/PageRankSuite.scala:324]: nit: `resetProb` is already a double type, it seems unnecessary to explicitly change it to 1.0, right?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorRollPlugin.scala
Comment (LuciferYang) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorRollPlugin.scala:154]: Directly change to `e.totalDuration.toFloat` seems to change the result? It might need @dongjoon-hyun  to confirm this change

File: streaming/src/main/scala/org/apache/spark/streaming/ui/UIUtils.scala
Comment (LuciferYang) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/UIUtils.scala:70]: How about unifying it to `milliseconds.toDouble`?

**Review by LuciferYang — COMMENTED** [review]
https://github.com/apache/spark/blob/398bff77b1c837aed55f4f10ff1157f7da813570/project/SparkBuild.scala#L239

The line above also needs to be deleted in this pr.

Additionally, you can provide an example in the pr description to clarify what the compile warning to be cleaned/fixed looks like.

File: core/src/main/scala/org/apache/spark/resource/ResourceUtils.scala
Comment (tgravescs) [reply core/src/main/scala/org/apache/spark/resource/ResourceUtils.scala:479]: it doesn't looks like the floor is needed here. its longs and ints, the other functions called before this take care of doubles.

Comment (hannahkamundson) [reply core/src/main/scala/org/apache/spark/resource/ResourceUtils.scala:479]: Changed it!

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorRollPlugin.scala
Comment (hannahkamundson) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorRollPlugin.scala:154]: Nice catch! Done.

File: graphx/src/test/scala/org/apache/spark/graphx/lib/PageRankSuite.scala
Comment (hannahkamundson) [reply graphx/src/test/scala/org/apache/spark/graphx/lib/PageRankSuite.scala:324]: Done!

File: streaming/src/main/scala/org/apache/spark/streaming/ui/UIUtils.scala
Comment (hannahkamundson) [reply streaming/src/main/scala/org/apache/spark/streaming/ui/UIUtils.scala:70]: Done! Let me know if my change isn't what you meant

File: core/src/test/scala/org/apache/spark/deploy/history/EventLogTestHelper.scala
Comment (hannahkamundson) [reply core/src/test/scala/org/apache/spark/deploy/history/EventLogTestHelper.scala:59]: Done!

File: common/unsafe/src/test/scala/org/apache/spark/unsafe/types/UTF8StringPropertyCheckSuite.scala
Comment (hannahkamundson) [reply common/unsafe/src/test/scala/org/apache/spark/unsafe/types/UTF8StringPropertyCheckSuite.scala:83]: No necessity! It is just what my IDE did when moving it to multiple lines. Is there a style preference in this codebase?

File: core/src/main/scala/org/apache/spark/input/FixedLengthBinaryInputFormat.scala
Comment (hannahkamundson) [reply core/src/main/scala/org/apache/spark/input/FixedLengthBinaryInputFormat.scala:77]: Done! Removed the `Math.floor(....)`

File: core/src/test/scala/org/apache/spark/benchmark/Benchmark.scala
Comment (hannahkamundson) [reply core/src/test/scala/org/apache/spark/benchmark/Benchmark.scala]: I think I'm understanding what you said. Let me know if you meant something different than I posted

**Comment by hannahkamundson** [comment]
Thanks for the review @LuciferYang . I fixed the changes and updated the PR. Feel free to give any more feedback :)

If it is okay, I will break up the second line [spark/project/SparkBuild.scala](https://github.com/apache/spark/blob/398bff77b1c837aed55f4f10ff1157f7da813570/project/SparkBuild.scala#L239) into another PR because I wasn't planning on that for this one. That'll make it easier for me to keep my changes straight.

**Comment by hannahkamundson** [comment]
Wait @LuciferYang I found this [Jira ticket 33805](https://issues.apache.org/jira/browse/SPARK-33805) for the [second line in the pom file](https://github.com/apache/spark/blob/25ee62e19bb00610e79de3284987970d9732195c/pom.xml#L2980). Are you sure I should do that for this Jira ticket? It seems like it may be for another ticket.

(I can help with that ticket too, but it helps me to keep them separate if possible.)

File: core/src/test/scala/org/apache/spark/benchmark/Benchmark.scala
Comment (LuciferYang) [reply core/src/test/scala/org/apache/spark/benchmark/Benchmark.scala]: Actually, I hope it's `val avg = runTimes.sum.toDouble / runTimes.size`, then `avg` will be more accurate.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorRollPlugin.scala
Comment (LuciferYang) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorRollPlugin.scala]: hmm... I'm not saying this change is wrong, but it needs confirmation from the developers of the k8s part.

File: core/src/main/scala/org/apache/spark/resource/ResourceUtils.scala
Comment (LuciferYang) [reply core/src/main/scala/org/apache/spark/resource/ResourceUtils.scala]: Thanks @tgravescs

**Comment by LuciferYang** [comment]
> Thanks for the review @LuciferYang . I fixed the changes and updated the PR. Feel free to give any more feedback :)
> 
> If it is okay, I will break up the second line [spark/project/SparkBuild.scala](https://github.com/apache/spark/blob/398bff77b1c837aed55f4f10ff1157f7da813570/project/SparkBuild.scala#L239) into another PR because I wasn't planning on that for this one. That'll make it easier for me to keep my changes straight.

We need to clean up the corresponding suppression rule in sbt in this pr to keep the suppression rules for Maven and sbt consistent. It doesn't need to be split into another PR.

**Comment by LuciferYang** [comment]
> Wait @LuciferYang I found this [Jira ticket 33805](https://issues.apache.org/jira/browse/SPARK-33805) for the [second line in the pom file](https://github.com/apache/spark/blob/25ee62e19bb00610e79de3284987970d9732195c/pom.xml#L2980). Are you sure I should do that for this Jira ticket? It seems like it may be for another ticket.
> 
> (I can help with that ticket too, but it helps me to keep them separate if possible.)

This part of the work has basically been completed. I will double-check it in the near future(After SPARK-45699 and SPARK-45629 are completed.), and if there are any new related tasks to be done, I will ping you in Jira.

**Comment by hannahkamundson** [comment]
> > Thanks for the review @LuciferYang . I fixed the changes and updated the PR. Feel free to give any more feedback :)
> > If it is okay, I will break up the second line [spark/project/SparkBuild.scala](https://github.com/apache/spark/blob/398bff77b1c837aed55f4f10ff1157f7da813570/project/SparkBuild.scala#L239) into another PR because I wasn't planning on that for this one. That'll make it easier for me to keep my changes straight.
> 
> We need to clean up the corresponding suppression rule in sbt in this pr to keep the suppression rules for Maven and sbt consistent. It doesn't need to be split into another PR.

Ohh. In response to my last comments, I was thinking you were saying the line above it in  Maven. I'll change the sbt one!

File: core/src/test/scala/org/apache/spark/benchmark/Benchmark.scala
Comment (hannahkamundson) [reply core/src/test/scala/org/apache/spark/benchmark/Benchmark.scala]: I will change it. Is it okay that we are changing the value of average (like you pointed out in the other one)?

Comment (hannahkamundson) [reply core/src/test/scala/org/apache/spark/benchmark/Benchmark.scala]: Done! Let me know if you want me to undo it so the value stays the same

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorRollPlugin.scala
Comment (hannahkamundson) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorRollPlugin.scala]: Oh ya makes sense! Right now I have it where the value will be the same as it originally was. Let me know if you want me to change that

File: project/SparkBuild.scala
Comment (LuciferYang) [inline project/SparkBuild.scala:239]: Sorry, I've reconsidered again. Let's not delete it and change it to 
```
-Wconf:msg=^(?=.*?Widening conversion from)(?=.*?is deprecated because it loses precision).+$:e
```

both sbt and maven

Comment (hannahkamundson) [reply project/SparkBuild.scala:239]: Hm, what is the goal of replacing it? My original intent was to get all compilation warnings that say `is deprecated because it loses precision`/`Widening conversion from` gone. It seems like this would be adding a compilation warning for something that isn't there. Thoughts?

Comment (hannahkamundson) [reply project/SparkBuild.scala:239]: Deleted my comment as I read it closer! I didn't see the changed `e`. Adding this in!

File: core/src/main/scala/org/apache/spark/partial/CountEvaluator.scala
Comment (LuciferYang) [inline core/src/main/scala/org/apache/spark/partial/CountEvaluator.scala:63]: nit:
```
new BoundedDouble(
  sum + dist.getNumericalMean, confidence, (sum + low).toDouble, (sum + high).toDouble)
```

File: core/src/main/scala/org/apache/spark/partial/GroupedCountEvaluator.scala
Comment (LuciferYang) [inline core/src/main/scala/org/apache/spark/partial/GroupedCountEvaluator.scala:47]: nit:
```scala
sums.map { case (key, sum) =>
  (key, new BoundedDouble(sum.toDouble, 1.0, sum.toDouble, sum.toDouble))
}.toMap
```

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala
Comment (LuciferYang) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:1209]:
```scala
foundTasks = checkAndSubmitSpeculatableTasks(
  timeMs, threshold.toDouble, customizedThreshold = true)
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala
Comment (LuciferYang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:1130]:
```scala
(upperIndex, upperBinValue,
  binAr :+ HistogramBin(lowerBinValue.toDouble, upperBinValue.toDouble, ndv))
```

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/EventTimeWatermarkSuite.scala
Comment (LuciferYang) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/EventTimeWatermarkSuite.scala:81]:
```scala
EventTimeStats(
  max = largeValue, min = largeValue, avg = largeValue.toDouble, count = largeValue - 1)
```

**Review by LuciferYang — COMMENTED** [review]
The code overall is fine to me, there are some minor formatting issues, I think unnecessary line breaks should be minimized.

File: core/src/main/scala/org/apache/spark/partial/CountEvaluator.scala
Comment (hannahkamundson) [reply core/src/main/scala/org/apache/spark/partial/CountEvaluator.scala:63]: Done!

File: core/src/main/scala/org/apache/spark/partial/GroupedCountEvaluator.scala
Comment (hannahkamundson) [reply core/src/main/scala/org/apache/spark/partial/GroupedCountEvaluator.scala:47]: Done!

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala
Comment (hannahkamundson) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:1209]: Done!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala
Comment (hannahkamundson) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:1130]: Done!

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/EventTimeWatermarkSuite.scala
Comment (hannahkamundson) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/EventTimeWatermarkSuite.scala:81]: Done!

**Comment by hannahkamundson** [comment]
@LuciferYang I updated the formatting! Let me know if you have any other comments before approval :)

**Review by LuciferYang — APPROVED** [review]
+1, LGTM

**Comment by LuciferYang** [comment]
image

It seems this is different from the current actual modifications, could you update the PR description?

**Comment by hannahkamundson** [comment]
@LuciferYang Updated! How do I merge it in?

**Comment by LuciferYang** [comment]
Merged into master for Spark 4.0. Thanks @hannahkamundson and @tgravescs ~

**Pull request closed without merging**
